src/PrettyPrinting.jl

```diff
@@ -1591,23 +1591,81 @@ function _write_line(io::IOCustom, str::AbstractString)
     spaceleft = c - ind - io.printed
   end
   #@show spaceleft
-  firstlen = min(spaceleft, length(str))
-  firststr = str[1:firstlen]
+  # The following code deals with line wrapping of Unicode text, including
+  # double-width symbols and more.
+  _graphemes = Base.Unicode.graphemes(str)
+  firstlen = min(spaceleft, length(_graphemes))
+  # make an iterator over valid indices
   if io.lowercasefirst
-    written += write(io.io, lowercasefirst(firststr))
+    firstiter = Base.Iterators.take(_graphemes, firstlen)
+    firstiter = join(firstiter)
+    firstiter = Base.Unicode.graphemes(lowercasefirst(firstiter))
     io.lowercasefirst = false
   else
-    written += write(io.io, firststr)
+    firstiter = Base.Iterators.take(_graphemes, firstlen)
     io.lowercasefirst = false
   end
-  io.printed += textwidth(firststr)
-  reststr = str[firstlen + 1:end]
-  it = Iterators.partition(1:textwidth(reststr), c - ind > 0 ? c - ind : c)
-  for i in it
+  restiter = Base.Iterators.drop(_graphemes, firstlen)
+  firststr = join(firstiter)
+  if length(firstiter) == textwidth(firststr)
+    written += write(io.io, firststr)
+    io.printed += textwidth(firststr)
```

Comment:
Computing `textwidth` is not free, so maybe remember the result?
```suggestion
  width = textwidth(firststr)
  if length(firstiter) == width
    written += write(io.io, firststr)
    io.printed += width
```
I guess would could also use `length(firststr)` (which I think at this point is equal to `length(firstiter)` and hence `textwidth(firststr)` but an explicit `width` seems clearer?)